odoo/service/server.py

```diff
@@ -742,11 +742,29 @@ def process_timeout(self):
         for (pid, worker) in self.workers.items():
             if worker.watchdog_timeout is not None and \
                     (now - worker.watchdog_time) >= worker.watchdog_timeout:
-                _logger.error("%s (%s) timeout after %ss",
-                              worker.__class__.__name__,
-                              pid,
-                              worker.watchdog_timeout)
-                self.worker_kill(pid, signal.SIGKILL)
+                if not worker.stack_dumped:
+                    _logger.error("%s (%s) timeout after %ss",
+                                  worker.__class__.__name__,
+                                  pid,
+                                  worker.watchdog_timeout)
+                    _logger.info("Giving %s (%s) %ss to dump stacktrace ",
+                                 worker.__class__.__name__,
+                                 pid,
+                                 self.beat)
+                    self.worker_kill(pid, signal.SIGQUIT)
+                    worker.stack_dumped = now
+                elif worker.stack_dumped and worker.stack_dumped > now - self.beat:
+                    pass
+                else:
+                    _logger.info("Killing %s (%s)",
+                                 worker.__class__.__name__,
+                                 pid)
+                    self.worker_kill(pid, signal.SIGKILL)
+            elif worker.stack_dumped:
+                _logger.info("%s (%s) just finished, keeping it alive",
```

Comment:
Is it true? `SIGQUIT` has already been sent.

Reply:
`SIGQUIT` only does a dumpstacks, I don't see a problem to keep the worker alive afterwards, do you?

Note that this cornercase is very unlikely to happen, the request would need to last exactly limit_time_real + [0-8] seconds.